### PR TITLE
DAOS-5435: weekly-testing: IorIntercept: test_ior_intercept: IOR command issued warnings

### PR DESCRIPTION
**Pull request by cibervicho**
DAOS-5435: weekly-testing: IorIntercept: test_ior_intercept: IOR command issued warnings

'ior' command uses the -D optional argument to specify the number of seconds
before stopping a read or write phase. If for some reason the data written or read
takes longer than the seconds allowed by this parameter the test logs WARNING
messages and finally fails the test.

Being exposed that, an optional boolean parameter named 'fail_on_warning' was introduced
in the run_ior_with_pool() function to control whether the test should fail or not.
Also added a DFS run in the test to have base performance numbers for comparison
purposes.

Signed-off-by: David Maldonado Moreno <david.maldonado.moreno@intel.com>

### CODE REVIEW

File: src/tests/ftest/io/ior_fail_on_warning.py
Comment (cibervicho) [reply src/tests/ftest/io/ior_fail_on_warning.py:55]: Done

Comment (cibervicho) [reply src/tests/ftest/io/ior_fail_on_warning.py:55]: Done

File: src/tests/ftest/io/ior_fail_on_warning.py
Comment (cibervicho) [reply src/tests/ftest/io/ior_fail_on_warning.py:53]: Done.

File: src/tests/ftest/io/ior_fail_on_warning.py
Comment (cibervicho) [reply src/tests/ftest/io/ior_fail_on_warning.py:28]: Done

**Review by saurabhtandan — CHANGES_REQUESTED** [review]
As per our conversation on MS Teams, we can do without this added test, as the changes made as part of this PR will be tested in ior_intercept_basic (the original failure).
In your next push, can you please add the following line to the final commit message to make sure ior_intercept_basic is run to verify the original bug is fixed.
`Test-tag-hw-small: pr,hw,small iorinterceptbasic`

File: src/tests/ftest/io/ior_intercept_basic.py
Comment (cibervicho) [reply src/tests/ftest/io/ior_intercept_basic.py:62]: Done

**Review by shimizukko — APPROVED** [review]
LGTM

**Comment by saurabhtandan** [comment]
The test impacted by this PR is failing in CI.

**Comment by cibervicho** [comment]
> The test impacted by this PR is failing in CI.

PR #3405 merged. I'm re-trying with this patch to see if there is an improvement

**Comment by cibervicho** [comment]
Added the "fail_on_warning" flag as requested. This flag is working as expected when WARNING messages are shown in log, however the test now fails with an assertion when comparing **POSIX read** vs **POSIX + IL read**.

When discussing this with Saurabh he suggested, after talking with Mohamad, to include a DFS run to the test to get some baseline numbers because the performance numbers with IL should be similar or closer to the numbers if we use DFS api.

I got the results here:
https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3570/16/testReport/junit/(root)/IorIntercept/Test___Functional_Hardware_Small___1___io_ior_intercept_basic_py_IorIntercept_test_ior_intercept_container_dfuse_hosts_np_16_oclass_SX_1M_pool_servers_ca08/

`
==== DFS ====
write   2629.70 MiB/sec
read    2795.45 MiB/sec

==== POSIX ====
without IL: write   329.81 MiB/sec
without IL: read    1383.25 MiB/sec
with IL:    write   2418.37 MiB/sec
with IL:    read    1339.22 MiB/sec
`

**DFS read > (POSIX + IL read = POSIX read)**

It seems to be a problem with the Intercept Library.

**Comment by cibervicho** [comment]
Opened the following ticket to address the IL bug: https://jira.hpdd.intel.com/browse/DAOS-5857

**Review by saurabhtandan — APPROVED** [review]
LGTM

**Review by shimizukko — APPROVED** [review]
LGTM

**Review by mchaarawi — CHANGES_REQUESTED** [review]
What are the warning messages that ior prints here? is it just the stat size warning?
what if there are other warnings?

I don't think you really need to do stonewalling here (-D 300). i would just remove that, and use a smaller block size, something like 4G and that should run in a reasonable amount of time. then you should not expect any warnings.

It would also be good to keep the DFS run before the runs over fuse to always have the comparison here.

**Comment by saurabhtandan** [comment]
> What are the warning messages that ior prints here? is it just the stat size warning?
> what if there are other warnings?
> 
> I don't think you really need to do stonewalling here (-D 300). i would just remove that, and use a smaller block size, something like 4G and that should run in a reasonable amount of time. then you should not expect any warnings.
> 
> It would also be good to keep the DFS run before the runs over fuse to always have the comparison here.

Yeah the warning here is due to stat size. If you want '-D' option to be removed for this test, i think it shouldn't be a problem.
But there are other tests which require ior to be run with '-D' option and are failing too in weekly run. They will need this option to avoid fail on warning. Without this option we cannot run ior with -D option in CI right now.

**Review by mchaarawi — APPROVED** [review]

**Comment by mchaarawi** [comment]
> > What are the warning messages that ior prints here? is it just the stat size warning?
> > what if there are other warnings?
> > I don't think you really need to do stonewalling here (-D 300). i would just remove that, and use a smaller block size, something like 4G and that should run in a reasonable amount of time. then you should not expect any warnings.
> > It would also be good to keep the DFS run before the runs over fuse to always have the comparison here.
> 
> Yeah the warning here is due to stat size. If you want '-D' option to be removed for this test, i think it shouldn't be a problem.
> But there are other tests which require ior to be run with '-D' option and are failing too in weekly run. They will need this option to avoid fail on warning. Without this option we cannot run ior with -D option in CI right now.

Ok I see. Thanks for the explanation. I’ll approve then if there are other requirements where stonewalling is needed.
But as I mentioned it would be good to add the DFS test here for comparison since I see it was added then reverted in the pr. It’s fine to do it in this or a separate pr.

**Comment by sylviachanoiyee** [comment]
Do you know why the iorintercept test failed in Jenkins?

**Comment by saurabhtandan** [comment]
> Do you know why the iorintercept test failed in Jenkins?

There is a parallel ticket opened for that issue.
https://jira.hpdd.intel.com/browse/DAOS-5857

**Comment by sylviachanoiyee** [comment]
@saurabhtandan , thanks.
@cibervicho , is the issue intermittent or consistent?
If it is consistently failing, please do the following:
1. add skipForTicket in this PR to skip the test for DAOS-5857). I will land it.
2. create a ticket in JIRA to re-enable this test and set it as blocking DAOS-5857.

Thanks!

**Review by saurabhtandan — DISMISSED** [review]
LGTM

File: src/tests/ftest/io/ior_intercept_basic.py
Comment (cibervicho) [reply src/tests/ftest/io/ior_intercept_basic.py:40]: Fixed.

**Review by shimizukko — DISMISSED** [review]
LGTM

**Comment by ashleypittman** [comment]
The -D option was to give us a vaguely consistent run-time but if our CI is stable now and running on consistent hardware we could probably remove it at the risks of tests taking longer.

The abort_on_warnings issue is something that I'd forgotten about, we really should land https://github.com/daos-stack/ior-hpc/commit/4ec2bebc33900c2712f21852eff685828d05f4c6 into IOR and then we wouldn't need this hack.

**Comment by saurabhtandan** [comment]
> The -D option was to give us a vaguely consistent run-time but if our CI is stable now and running on consistent hardware we could probably remove it at the risks of tests taking longer.
> 
> The abort_on_warnings issue is something that I'd forgotten about, we really should land [daos-stack/ior-hpc@4ec2beb](https://github.com/daos-stack/ior-hpc/commit/4ec2bebc33900c2712f21852eff685828d05f4c6) into IOR and then we wouldn't need this hack.

stonewalling is a handy option I think. Otherwise we will be playing a guessing game with time limits everything in test side. Anyways, that's great if the above PR will take care of this issue. Until it is merged, it will be nice to go ahead with this hack. The PR seems almost ready, given the current run goes through fine.

**Comment by cibervicho** [comment]
> @saurabhtandan , thanks.
> @cibervicho , is the issue intermittent or consistent?
> If it is consistently failing, please do the following:
> 
> 1. add skipForTicket in this PR to skip the test for DAOS-5857). I will land it.
> 2. create a ticket in JIRA to re-enable this test and set it as blocking DAOS-5857.
> 
> Thanks!

@sylviachanoiyee I completed step1 above. The test was skipped in CI.

**Review by shimizukko — APPROVED** [review]
LGTM

**Comment by cibervicho** [comment]
> > Do you know why the iorintercept test failed in Jenkins?
> 
> There is a parallel ticket opened for that issue.
> https://jira.hpdd.intel.com/browse/DAOS-5857

Will merge with master and send a new patch to double check the fix.

**Comment by cibervicho** [comment]
> > > Do you know why the iorintercept test failed in Jenkins?
> > 
> > 
> > There is a parallel ticket opened for that issue.
> > https://jira.hpdd.intel.com/browse/DAOS-5857
> 
> Will merge with master and send a new patch to double check the fix.

These are the results of last night's CI run (ior_intercept_basic.py):

==== DFS ====
without IL: write   2765.40 MiB/sec
without IL: read    2691.96 MiB/sec

==== POSIX ====
without IL: write   317.69 MiB/sec
without IL: read    1387.19 MiB/sec
with IL:    write   2465.06
with IL:    read    1350.48

The test also failed with the "AssertionError: False is not true" error:

06:34:15 ERROR| Reproduced traceback from: /usr/lib/python2.7/site-packages/avocado/core/test.py:596
06:34:15 ERROR| Traceback (most recent call last):
06:34:15 ERROR|   File "/usr/lib/daos/TESTING/ftest/io/ior_intercept_basic.py", line 87, in test_ior_intercept
06:34:15 ERROR|     read_x * float(without_intercept[1][max_mib]))
06:34:15 ERROR|   File "/usr/lib64/python2.7/unittest/case.py", line 462, in assertTrue
06:34:15 ERROR|     raise self.failureException(msg)
06:34:15 ERROR| AssertionError: False is not true

https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-3570/21/testReport/(root)/IorIntercept/Test___Functional_Hardware_Small___1___io_ior_intercept_basic_py_IorIntercept_test_ior_intercept_container_dfuse_hosts_np_16_oclass_SX_1M_pool_servers_74f0/

I'm readding the skipForTicket decorator to the test and re-submitting to CI.

**Comment by cibervicho** [comment]
All CI checks passed.

**Review by saurabhtandan — APPROVED** [review]
LGTM

**Review by sylviachanoiyee — APPROVED** [review]

**Comment by sylviachanoiyee** [comment]
@cibervicho , please create a JIRA ticket to re-enable this test and set it to block DAOS-5857. thanks!

**Pull request merged**
